<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ резюме из HeadHunter
   

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

# Исследование структуры данных

1. Прочитайте данные с помощью библиотеки Pandas. Совет: перед чтением обратите внимание на разделитель внутри файла. 

In [6]:
df = pd.read_csv(r'/Users/seegment/Desktop/Магистратура/Программирование на Python/Домашнее задание 5/dst-3.0_16_1_hh_database.csv', sep = ';', encoding='utf-8') #ваш код здесь


In [7]:
df

Пол, возраст           ЗП  \
0        Мужчина ,  39 лет , родился 27 ноября 1979   29000 руб.   
1         Мужчина ,  60 лет , родился 20 марта 1959   40000 руб.   
2      Женщина ,  36 лет , родилась 12 августа 1982   20000 руб.   
3          Мужчина ,  38 лет , родился 25 июня 1980  100000 руб.   
4         Женщина ,  26 лет , родилась 3 марта 1993  140000 руб.   
...                                             ...          ...   
44739    Мужчина ,  30 лет , родился 17 января 1989   50000 руб.   
44740      Мужчина ,  27 лет , родился 5 марта 1992   39000 руб.   
44741  Женщина ,  48 лет , родилась 26 декабря 1970   40000 руб.   
44742   Мужчина ,  24 года , родился 6 октября 1994   20000 руб.   
44743    Мужчина ,  38 лет , родился 25 апреля 1980  120000 руб.   

                               Ищет работу на должность:  \
0                                Системный администратор   
1                                   Технический писатель   
2                                               Оператор   
3      Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4                      Региональный менеджер по продажам   
...                                                  ...   
44739  Финансист, аналитик, экономист, бухгалтер, мен...   
44740             Системный администратор, IT-специалист   
44741                         Аналитик данных, Математик   
44742                                   Контент-менеджер   
44743                               Руководитель проекта   

                            Город, переезд, командировки  \
0      Советск (Калининградская область) , не готов к...   
1      Королев , не готов к переезду , готов к редким...   
2      Тверь , не готова к переезду , не готова к ком...   
3      Саратов , не готов к переезду , готов к редким...   
4      Москва , не готова к переезду , готова к коман...   
...                                                  ...   
44739  Тверь , готов к переезду (Москва, Химки) , гот...   
44740  Липецк , готов к переезду , готов к командировкам   
44741  Челябинск , готова к переезду , готова к редки...   
44742  Тамбов , не готов к переезду , не готов к кома...   
44743  Москва , не готов к переезду , не готов к кома...   

                                               Занятость  \
0      частичная занятость, проектная работа, полная ...   
1      частичная занятость, проектная работа, полная ...   
2                                       полная занятость   
3      частичная занятость, проектная работа, полная ...   
4                                       полная занятость   
...                                                  ...   
44739                                   полная занятость   
44740  проектная работа, частичная занятость, полная ...   
44741                                   полная занятость   
44742              частичная занятость, полная занятость   
44743                                   полная занятость   

                                                  График  \
0      гибкий график, полный день, сменный график, ва...   
1      гибкий график, полный день, сменный график, уд...   
2                                            полный день   
3                        гибкий график, удаленная работа   
4                                            полный день   
...                                                  ...   
44739                      полный день, удаленная работа   
44740  удаленная работа, гибкий график, полный день, ...   
44741                      полный день, удаленная работа   
44742                                   удаленная работа   
44743                                        полный день   

                                             Опыт работы  \
0      Опыт работы 16 лет 10 месяцев  Август 2010 — п...   
1      Опыт работы 19 лет 5 месяцев  Январь 2000 — по...   
2      Опыт работы 10 лет 3 месяца  Октябрь 2004 — Де...   
3      Опыт работы 18 лет 9 месяцев  Август 2017 — Ап...   
4      Опыт работы 5 лет 7 месяцев  Ре

2. Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.

In [9]:
# Вывести несколько первых строк таблицы
print("Первые строки таблицы:")
print(df.head(2))  # Вывод первых 5 строк

# Вывести несколько последних строк таблицы
print("\nПоследние строки таблицы:")
print(df.tail(2))  # Вывод последних 5 строк

Первые строки таблицы:
                                 Пол, возраст          ЗП  \
0  Мужчина ,  39 лет , родился 27 ноября 1979  29000 руб.   
1   Мужчина ,  60 лет , родился 20 марта 1959  40000 руб.   

  Ищет работу на должность:  \
0   Системный администратор   
1      Технический писатель   

                        Город, переезд, командировки  \
0  Советск (Калининградская область) , не готов к...   
1  Королев , не готов к переезду , готов к редким...   

                                           Занятость  \
0  частичная занятость, проектная работа, полная ...   
1  частичная занятость, проектная работа, полная ...   

                                              График  \
0  гибкий график, полный день, сменный график, ва...   
1  гибкий график, полный день, сменный график, уд...   

                                         Опыт работы  \
0  Опыт работы 16 лет 10 месяцев  Август 2010 — п...   
1  Опыт работы 19 лет 5 месяцев  Январь 2000 — по...   

  Последнее/нынешнее ме

3. Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.

In [11]:
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

4. Обратите внимание на информацию о числе непустых значений.

In [13]:
missing_values = df.isnull().sum()#ваш код здесь
missing_values

Пол, возраст                         0
ЗП                                   0
Ищет работу на должность:            0
Город, переезд, командировки         0
Занятость                            0
График                               0
Опыт работы                        168
Последнее/нынешнее место работы      1
Последняя/нынешняя должность         2
Образование и ВУЗ                    0
Обновление резюме                    0
Авто                                 0
dtype: int64

Согласно анализу данных, отсутствующие значения (пропуски) наблюдаются в следующих столбцах:

**Опыт работы** — 168 пропусков.  
**Последнее/нынешнее место работы** — 1 пропуск.  
**Последняя/нынешняя должность** — 2 пропуска.  

Вывод:
Большинство столбцов в таблице заполнены полностью, что свидетельствует о хорошем качестве данных.  
Однако в трёх столбцах есть пропущенные значения:  
В столбце "Опыт работы" значительное количество пропусков (168). Это может быть связано с тем, что некоторые респонденты не указали свой опыт.  
В столбцах "Последнее/нынешнее место работы" и "Последняя/нынешняя должность" пропусков меньше (1 и 2 соответственно), что менее критично.  

Для столбца **"Опыт работы"**:
Определить, можно ли заполнить пропуски средним, медианой или каким-то другим логичным значением.
Если опыт работы обязателен, можно исключить строки с пропусками.

Для столбцов **"Последнее/нынешнее место работы"** и **"Последняя/нынешняя должность"**:
Изучить строки с пропусками. Возможно, их можно заполнить вручную или исключить.

5. Выведите основную статистическую информацию о столбцах.


In [16]:
# Статистическая информация обо всех столбцах (включая текстовые)
print("\nОсновная статистическая информация обо всех столбцах:")
print(df.describe())#ваш код здесь


Основная статистическая информация обо всех столбцах:
                                         Пол, возраст          ЗП  \
count                                           44744       44744   
unique                                          16003         690   
top     Мужчина ,  32 года , родился 17 сентября 1986  50000 руб.   
freq                                               18        4064   

       Ищет работу на должность:  \
count                      44744   
unique                     14929   
top      Системный администратор   
freq                        3099   

                             Город, переезд, командировки         Занятость  \
count                                               44744             44744   
unique                                              10063                38   
top     Москва , не готов к переезду , не готов к кома...  полная занятость   
freq                                                 1261             30026   

             График   

# Преобразование данных

1. Начнем с простого - с признака **"Образование и ВУЗ"**. Его текущий формат это: **<Уровень образования год выпуска ВУЗ специальность...>**. Например:
* Высшее образование 2016 Московский авиационный институт (национальный исследовательский университет)...
* Неоконченное высшее образование 2000  Балтийская государственная академия рыбопромыслового флота…
Нас будет интересовать только уровень образования.

Создайте с помощью функции-преобразования новый признак **"Образование"**, который должен иметь 4 категории: "высшее", "неоконченное высшее", "среднее специальное" и "среднее".

Выполните преобразование, ответьте на контрольные вопросы и удалите признак "Образование и ВУЗ".

Совет: обратите внимание на структуру текста в столбце **"Образование и ВУЗ"**. Гарантируется, что текущий уровень образования соискателя всегда находится в первых 2ух слов и начинается с заглавной буквы. Воспользуйтесь этим.

*Совет: проверяйте полученные категории, например, с помощью метода unique()*


In [19]:
# Функция для извлечения уровня образования
def extract_education_level(text):
    if text.startswith("Высшее"):
        return "высшее"
    elif text.startswith("Неоконченное высшее"):
        return "неоконченное высшее"
    elif text.startswith("Среднее специальное"):
        return "среднее специальное"
    elif text.startswith("Среднее"):
        return "среднее"
    else:
        return None  # На случай некорректных данных

# Создание нового признака "Образование"
df['Образование'] = df['Образование и ВУЗ'].apply(extract_education_level)

# Проверка уникальных значений в новом признаке
print("Уникальные значения в признаке 'Образование':")
print(df['Образование'].unique())

# Удаление столбца "Образование и ВУЗ"
df = df.drop(columns=['Образование и ВУЗ'])

# Вывод первых строк для проверки
print("\nПервые строки данных после преобразования:")
print(df.head())


Уникальные значения в признаке 'Образование':
['неоконченное высшее' 'высшее' 'среднее специальное' 'среднее']

Первые строки данных после преобразования:
                                   Пол, возраст           ЗП  \
0    Мужчина ,  39 лет , родился 27 ноября 1979   29000 руб.   
1     Мужчина ,  60 лет , родился 20 марта 1959   40000 руб.   
2  Женщина ,  36 лет , родилась 12 августа 1982   20000 руб.   
3      Мужчина ,  38 лет , родился 25 июня 1980  100000 руб.   
4     Женщина ,  26 лет , родилась 3 марта 1993  140000 руб.   

                           Ищет работу на должность:  \
0                            Системный администратор   
1                               Технический писатель   
2                                           Оператор   
3  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4                  Региональный менеджер по продажам   

                        Город, переезд, командировки  \
0  Советск (Калининградская область) , не готов к...   
1  Королев 

2. Теперь нас интересует столбец **"Пол, возраст"**. Сейчас он представлен в формате **<Пол , возраст , дата рождения >**. Например:
* Мужчина , 39 лет , родился 27 ноября 1979 
* Женщина , 21 год , родилась 13 января 2000
Как вы понимаете, нам необходимо выделить каждый параметр в отдельный столбец.

Создайте два новых признака **"Пол"** и **"Возраст"**. При этом важно учесть:
* Признак пола должен иметь 2 уникальных строковых значения: 'М' - мужчина, 'Ж' - женщина. 
* Признак возраста должен быть представлен целыми числами.

Выполните преобразование, ответьте на контрольные вопросы и удалите признак **"Пол, возраст"** из таблицы.

*Совет: обратите внимание на структуру текста в столбце, в части на то, как разделены параметры пола, возраста и даты рождения между собой - символом ' , '. 
Гарантируется, что структура одинакова для всех строк в таблице. Вы можете воспользоваться этим.*


In [21]:
import re

# Функция для извлечения пола
def extract_gender(text):
    # Убираем лишние пробелы и разделяем строку по символу ','
    parts = text.strip().split(' , ')
    # Извлекаем пол
    if 'Мужчина' in parts[0]:
        return 'М'
    elif 'Женщина' in parts[0]:
        return 'Ж'
    return None  # На случай некорректных данных

# Функция для извлечения возраста
def extract_age(text):
    try:
        # Убираем лишние пробелы и разделяем строку по символу ','
        parts = text.strip().split(' , ')
        # Используем регулярное выражение для извлечения возраста
        age_match = re.search(r'(\d+)\s*(лет|год|года)', parts[1])
        if age_match:
            return int(age_match.group(1))  # Возвращаем возраст как целое число
        return None  # Если возраст не найден, возвращаем None
    except (IndexError, ValueError):
        return None  # Если возникла ошибка, возвращаем None

# Создание нового признака "Пол"
df['Пол'] = df['Пол, возраст'].apply(extract_gender)

# Создание нового признака "Возраст"
df['Возраст'] = df['Пол, возраст'].apply(extract_age)

# Проверка уникальных значений в новых признаках
print("Уникальные значения в признаке 'Пол':")
print(df['Пол'].unique())

print("\nУникальные значения в признаке 'Возраст':")
print(df['Возраст'].unique())

# Удаление столбца "Пол, возраст"
df = df.drop(columns=['Пол, возраст'])

# Вывод первых строк для проверки
print("\nПервые строки данных после преобразования:")
print(df.head())


Уникальные значения в признаке 'Пол':
['М' 'Ж']

Уникальные значения в признаке 'Возраст':
[ 39  60  36  38  26  29  46  34  35  59  27  21  31  37  30  33  42  32
  23  28  25  24  43  53  49  44  48  54  22  20  40  62  72  45  41  61
  50  58  57  52  56  18  19  51  16  47  71  63  55  64  68  17  73  66
  65  77  67  69  76  70  14  15 100]

Первые строки данных после преобразования:
            ЗП                          Ищет работу на должность:  \
0   29000 руб.                            Системный администратор   
1   40000 руб.                               Технический писатель   
2   20000 руб.                                           Оператор   
3  100000 руб.  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4  140000 руб.                  Региональный менеджер по продажам   

                        Город, переезд, командировки  \
0  Советск (Калининградская область) , не готов к...   
1  Королев , не готов к переезду , готов к редким...   
2  Тверь , не готова к пе

3. Следующим этапом преобразуем признак **"Опыт работы"**. Его текущий формат - это: **<Опыт работы: n лет m месяцев, периоды работы в различных компаниях…>**. 

Из столбца нам необходимо выделить общий опыт работы соискателя в месяцах, новый признак назовем "Опыт работы (месяц)"

Для начала обсудим условия решения задачи:
* Во-первых, в данном признаке есть пропуски. Условимся, что если мы встречаем пропуск, оставляем его как есть (функция-преобразование возвращает NaN)
* Во-вторых, в данном признаке есть скрытые пропуски. Для некоторых соискателей в столбце стоит значения "Не указано". Их тоже обозначим как NaN (функция-преобразование возвращает NaN)
* В-третьих, нас не интересует информация, которая описывается после указания опыта работы (периоды работы в различных компаниях)
* В-четвертых, у нас есть проблема: опыт работы может быть представлен только в годах или только месяцах. Например, можно встретить следующие варианты:
    * Опыт работы 3 года 2 месяца…
    * Опыт работы 4 года…
    * Опыт работы 11 месяцев…
    * Учитывайте эту особенность в вашем коде

Учитывайте эту особенность в вашем коде

В результате преобразования у вас должен получиться столбец, содержащий информацию о том, сколько месяцев проработал соискатель.
Выполните преобразование, ответьте на контрольные вопросы и удалите столбец **"Опыт работы"** из таблицы.


In [23]:
# Создаём функцию для преобразования опыта работы
def calculate_experience_in_months(experience_string):
    # Проверяем на пропуски или скрытые пропуски
    if pd.isna(experience_string) or experience_string == "Не указано":
        return np.nan
    
    # Извлекаем значения лет и месяцев с использованием регулярных выражений
    import re
    years_match = re.search(r"(\d+)\s+лет", experience_string)
    months_match = re.search(r"(\d+)\s+месяц", experience_string)

    # Присваиваем значение лет и месяцев, если они найдены
    years = int(years_match.group(1)) if years_match else 0
    months = int(months_match.group(1)) if months_match else 0

    # Переводим общий опыт работы в месяцы
    total_months = years * 12 + months
    return total_months

# Применяем функцию к столбцу "Опыт работы" и создаём новый столбец "Опыт работы (месяц)"
df['Опыт работы (месяц)'] = df['Опыт работы'].apply(calculate_experience_in_months)

# Проверка уникальных значений в новом признаке
print(df['Опыт работы (месяц)'].value_counts())

# Удаление столбца "Опыт работы"
df = df.drop(columns=['Опыт работы'])

# Вывод первых строк для проверки
print("\nПервые строки данных после преобразования:")
print(df.head())#ваш код здесь



Опыт работы (месяц)
9.0      1254
8.0      1216
10.0     1134
5.0      1115
4.0      1082
         ... 
556.0       1
32.0        1
33.0        1
483.0       1
312.0       1
Name: count, Length: 372, dtype: int64

Первые строки данных после преобразования:
            ЗП                          Ищет работу на должность:  \
0   29000 руб.                            Системный администратор   
1   40000 руб.                               Технический писатель   
2   20000 руб.                                           Оператор   
3  100000 руб.  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4  140000 руб.                  Региональный менеджер по продажам   

                        Город, переезд, командировки  \
0  Советск (Калининградская область) , не готов к...   
1  Королев , не готов к переезду , готов к редким...   
2  Тверь , не готова к переезду , не готова к ком...   
3  Саратов , не готов к переезду , готов к редким...   
4  Москва , не готова к переезду , готова к кома

In [24]:
print(df['Опыт работы (месяц)'].value_counts().sort_index(ascending=False))

Опыт работы (месяц)
1203.0       1
1191.0       1
663.0        1
603.0        1
586.0        1
          ... 
4.0       1082
3.0       1019
2.0        894
1.0        846
0.0        281
Name: count, Length: 372, dtype: int64


4. Хорошо идем! Следующий на очереди признак "Город, переезд, командировки". Информация в нем представлена в следующем виде: **<Город , (метро) , готовность к переезду (города для переезда) , готовность к командировкам>**. В скобках указаны необязательные параметры строки. Например, можно встретить следующие варианты:

* Москва , не готов к переезду , готов к командировкам
* Москва , м. Беломорская , не готов к переезду, не готов к командировкам
* Воронеж , готов к переезду (Сочи, Москва, Санкт-Петербург) , готов к командировкам

Создадим отдельные признаки **"Город"**, **"Готовность к переезду"**, **"Готовность к командировкам"**. При этом важно учесть:

* Признак **"Город"** должен содержать только 4 категории: "Москва", "Санкт-Петербург" и "город-миллионник" (их список ниже), остальные обозначьте как "другие".

    Список городов-миллионников:
    
   <code>million_cities = ['Новосибирск', 'Екатеринбург','Нижний Новгород','Казань', 'Челябинск','Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж','Волгоград']
    </code>
    Инфорация о метро, рядом с которым проживает соискатель нас не интересует.
* Признак **"Готовность к переезду"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к переезду в признаке "Город, переезд, командировки". Например:
    * … , готов к переезду , …
    * … , не готова к переезду , …
    * … , готова к переезду (Москва, Санкт-Петербург, Ростов-на-Дону)
    * … , хочу переехать (США) , …
    
    Нас интересует только сам факт возможности или желания переезда.
* Признак **"Готовность к командировкам"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к командировкам в признаке "Город, переезд, командировки". Например:
    * … , готов к командировкам , … 
    * … , готова к редким командировкам , …
    * … , не готов к командировкам , …
    
    Нас интересует только сам факт готовности к командировке.
    
    Еще один важный факт: при выгрузки данных у некоторых соискателей "потерялась" информация о готовности к командировкам. Давайте по умолчанию будем считать, что такие соискатели не готовы к командировкам.
    
Выполните преобразования и удалите столбец **"Город, переезд, командировки"** из таблицы.

*Совет: обратите внимание на то, что структура текста может меняться в зависимости от указания ближайшего метро. Учите это, если будете использовать порядок слов в своей программе.*


In [26]:
# Список городов-миллионников
million_cities = [
    'Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань', 'Челябинск',
    'Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь',
    'Воронеж', 'Волгоград'
]

# Функция для определения города
def parse_city(city_str):
    if "Москва" in city_str:
        return "Москва"
    elif "Санкт-Петербург" in city_str:
        return "Санкт-Петербург"
    elif any(city in city_str for city in million_cities):
        return "город-миллионник"
    else:
        return "другие"

# Функция для определения готовности к переезду
def parse_relocation(city_str):
    if "готов к переезду" in city_str or "готова к переезду" in city_str or "хочу переехать" in city_str:
        return True
    else:
        return False

# Функция для определения готовности к командировкам
def parse_travel(city_str):
    if "готов к командировкам" in city_str or "готова к командировкам" in city_str:
        return True
    elif "не готов к командировкам" in city_str or "не готова к командировкам" in city_str:
        return False
    else:
        return False  # Для случаев, где "потерялась" информация

# Преобразуем признаки в DataFrame
df['Город'] = df['Город, переезд, командировки'].apply(parse_city)
df['Готовность к переезду'] = df['Город, переезд, командировки'].apply(parse_relocation)
df['Готовность к командировкам'] = df['Город, переезд, командировки'].apply(parse_travel)

# Удаляем старый столбец
df = df.drop(columns=['Город, переезд, командировки'])

# Результат
print(df.head())
#ваш код здесь

            ЗП                          Ищет работу на должность:  \
0   29000 руб.                            Системный администратор   
1   40000 руб.                               Технический писатель   
2   20000 руб.                                           Оператор   
3  100000 руб.  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4  140000 руб.                  Региональный менеджер по продажам   

                                           Занятость  \
0  частичная занятость, проектная работа, полная ...   
1  частичная занятость, проектная работа, полная ...   
2                                   полная занятость   
3  частичная занятость, проектная работа, полная ...   
4                                   полная занятость   

                                              График  \
0  гибкий график, полный день, сменный график, ва...   
1  гибкий график, полный день, сменный график, уд...   
2                                        полный день   
3                    гиб

5. Рассмотрим поближе признаки **"Занятость"** и **"График"**. Сейчас признаки представляют собой набор категорий желаемой занятости (полная занятость, частичная занятость, проектная работа, волонтерство, стажировка) и желаемого графика работы (полный день, сменный график, гибкий график, удаленная работа, вахтовый метод).
На сайте hh.ru соискатель может указывать различные комбинации данных категорий, например:
* полная занятость, частичная занятость
* частичная занятость, проектная работа, волонтерство
* полный день, удаленная работа
* вахтовый метод, гибкий график, удаленная работа, полная занятость

Такой вариант признаков имеет множество различных комбинаций, а значит множество уникальных значений, что мешает анализу. Нужно это исправить!

Давайте создадим признаки-мигалки для каждой категории: если категория присутствует в списке желаемых соискателем, то в столбце на месте строки рассматриваемого соискателя ставится True, иначе - False.

Такой метод преобразования категориальных признаков называется One Hot Encoding и его схема представлена на рисунке ниже:
<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/ohe.jpg>
Выполните данное преобразование для признаков "Занятость" и "График", ответьте на контрольные вопросы, после чего удалите их из таблицы

In [28]:
# Преобразуем "Занятость" и "График" с помощью One Hot Encoding
employment_dummies = df["Занятость"].str.get_dummies(sep=", ")
schedule_dummies = df["График"].str.get_dummies(sep=", ")

# Объединяем с исходным DataFrame
df = pd.concat([df, employment_dummies, schedule_dummies], axis=1)

# Удаляем оригинальные столбцы
df = df.drop(columns=["Занятость", "График"])

print(df)

                ЗП                          Ищет работу на должность:  \
0       29000 руб.                            Системный администратор   
1       40000 руб.                               Технический писатель   
2       20000 руб.                                           Оператор   
3      100000 руб.  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4      140000 руб.                  Региональный менеджер по продажам   
...            ...                                                ...   
44739   50000 руб.  Финансист, аналитик, экономист, бухгалтер, мен...   
44740   39000 руб.             Системный администратор, IT-специалист   
44741   40000 руб.                         Аналитик данных, Математик   
44742   20000 руб.                                   Контент-менеджер   
44743  120000 руб.                               Руководитель проекта   

                         Последнее/нынешнее место работы  \
0                                МАОУ "СОШ № 1 г.Немана"   
1  

6. (2 балла) Наконец, мы добрались до самого главного и самого важного - признака заработной платы **"ЗП"**. 
В чем наша беда? В том, что помимо желаемой заработной платы соискатель указывает валюту, в которой он бы хотел ее получать, например:
* 30000 руб.
* 50000 грн.
* 550 USD

Нам бы хотелось видеть заработную плату в единой валюте, например, в рублях. Возникает вопрос, а где взять курс валют по отношению к рублю?

На самом деле язык Python имеет в арсенале огромное количество возможностей получения данной информации, от обращения к API Центробанка, до использования специальных библиотек, например pycbrf. Однако, это не тема нашего проекта.

Поэтому мы пойдем в лоб: обратимся к специальным интернет-ресурсам для получения данных о курсе в виде текстовых файлов. Например, MDF.RU, данный ресурс позволяет удобно экспортировать данные о курсах различных валют и акций за указанные периоды в виде csv файлов. Мы уже сделали выгрузку курсов валют, которые встречаются в наших данных за период с 29.12.2017 по 05.12.2019. Скачать ее вы можете **на платформе**

Создайте новый DataFrame из полученного файла. В полученной таблице нас будут интересовать столбцы:
* "currency" - наименование валюты в ISO кодировке,
* "date" - дата, 
* "proportion" - пропорция, 
* "close" - цена закрытия (последний зафиксированный курс валюты на указанный день).


Перед вами таблица соответствия наименований иностранных валют в наших данных и их общепринятых сокращений, которые представлены в нашем файле с курсами валют. Пропорция - это число, за сколько единиц валюты указан курс в таблице с курсами. Например, для казахстанского тенге курс на 20.08.2019 составляет 17.197 руб. за 100 тенге, тогда итоговый курс равен - 17.197 / 100 = 0.17197 руб за 1 тенге.
Воспользуйтесь этой информацией в ваших преобразованиях.

<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/table.jpg>


Осталось только понять, откуда брать дату, по которой определяется курс? А вот же она - в признаке **"Обновление резюме"**, в нем содержится дата и время, когда соискатель выложил текущий вариант своего резюме. Нас интересует только дата, по ней бы и будем сопоставлять курсы валют.

Теперь у нас есть вся необходимая информация для того, чтобы создать признак "ЗП (руб)" - заработная плата в рублях.

После ответа на контрольные вопросы удалите исходный столбец заработной платы "ЗП" и все промежуточные столбцы, если вы их создавали.

Итак, давайте обсудим возможный алгоритм преобразования: 
1. Перевести признак "Обновление резюме" из таблицы с резюме в формат datetime и достать из него дату. В тот же формат привести признак "date" из таблицы с валютами.
2. Выделить из столбца "ЗП" сумму желаемой заработной платы и наименование валюты, в которой она исчисляется. Наименование валюты перевести в стандарт ISO согласно с таблицей выше.
3. Присоединить к таблице с резюме таблицу с курсами по столбцам с датой и названием валюты (подумайте, какой тип объединения надо выбрать, чтобы в таблице с резюме сохранились данные о заработной плате, изначально представленной в рублях). Значение close для рубля заполнить единицей 1 (курс рубля самого к себе)
4. Умножить сумму желаемой заработной платы на присоединенный курс валюты (close) и разделить на пропорцию (обратите внимание на пропуски после объединения в этих столбцах), результат занести в новый столбец "ЗП (руб)".


In [31]:
currency_rates = pd.read_csv('C:/Users/user/Downloads/rates.csv', sep = ',', encoding='utf-8')
currency_rates['date'] = pd.to_datetime(currency_rates['date'], format='%d/%m/%y')
print(currency_rates)
df['Дата'] = pd.to_datetime(df['Обновление резюме'], format='%d.%m.%Y %H:%M').dt.date
df['Дата'] = pd.to_datetime(df['Дата'], format='%d/%m/%y')

     currency per       date   time    close  vol  proportion
0         USD   D 2017-12-29  00:00  57.6291    0           1
1         USD   D 2017-12-30  00:00  57.6002    0           1
2         USD   D 2017-12-31  00:00  57.6002    0           1
3         USD   D 2018-01-01  00:00  57.6002    0           1
4         USD   D 2018-01-02  00:00  57.6002    0           1
...       ...  ..        ...    ...      ...  ...         ...
5659      AZN   D 2019-12-02  00:00  37.7729    0           1
5660      AZN   D 2019-12-03  00:00  37.9662    0           1
5661      AZN   D 2019-12-04  00:00  37.8073    0           1
5662      AZN   D 2019-12-05  00:00  37.8396    0           1
5663      AZN   D 2019-12-06  00:00  37.6148    0           1

[5664 rows x 7 columns]


In [32]:
import re

# Маппинг валют на ISO и пропорцию
currency_mapping = {
    "руб.": ("RUB", 1),
    "USD": ("USD", 1),
    "грн.": ("UAH", 10),
    "EUR": ("EUR", 1),
    "бел.руб.": ("BYN", 1),
    "KGS": ("KGS", 10),
    "сум": ("UZS", 10000),
    "AZN": ("AZN", 1),
    "KZT": ("KZT", 100)
}

# Выделяем сумму и валюту из столбца "ЗП"
df[["Сумма", "Валюта"]] = df["ЗП"].str.split(" ", expand=True)
df["Сумма"] = df["Сумма"].astype(float)

print(df)

                ЗП                          Ищет работу на должность:  \
0       29000 руб.                            Системный администратор   
1       40000 руб.                               Технический писатель   
2       20000 руб.                                           Оператор   
3      100000 руб.  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4      140000 руб.                  Региональный менеджер по продажам   
...            ...                                                ...   
44739   50000 руб.  Финансист, аналитик, экономист, бухгалтер, мен...   
44740   39000 руб.             Системный администратор, IT-специалист   
44741   40000 руб.                         Аналитик данных, Математик   
44742   20000 руб.                                   Контент-менеджер   
44743  120000 руб.                               Руководитель проекта   

                         Последнее/нынешнее место работы  \
0                                МАОУ "СОШ № 1 г.Немана"   
1  

In [33]:
# Преобразование валют в ISO и добавление пропорции
df[["Валюта", "Пропорция"]] = df["Валюта"].apply(
    lambda x: pd.Series(currency_mapping.get(x, (None, None)))
)

# Объединение с таблицей курсов валют
df = df.merge(
    currency_rates,
    left_on=["Дата", "Валюта"],
    right_on=["date", "currency"],
    how="left"  # Left join сохраняет все строки из таблицы резюме
)

# Заполнение пропусков курса для RUB
df.loc[df["Валюта"] == "RUB", ["close", "proportion"]] = [1, 1]

# Пересчёт ЗП в рублях
df["ЗП (руб)"] = (df["Сумма"] * df["close"]) / df["Пропорция"]

df = df.drop(columns=["Пропорция", "currency", "per", "date", "time", "close", "vol", "proportion", 'Валюта', 'Сумма'], errors="ignore")

In [34]:
df = df.drop(columns=["Пропорция", "currency", "per", "date", "time", "close", "vol", "proportion"], errors="ignore")
print(df.head(40))

               ЗП                          Ищет работу на должность:  \
0      29000 руб.                            Системный администратор   
1      40000 руб.                               Технический писатель   
2      20000 руб.                                           Оператор   
3     100000 руб.  Веб-разработчик (HTML / CSS / JS / PHP / базы ...   
4     140000 руб.                  Региональный менеджер по продажам   
5      25000 руб.                             Технический специалист   
6     250000 руб.                           Руководитель ИТ-проектов   
7      70000 руб.                                     Инженер АСУ ТП   
8      65000 руб.                                            Ревизор   
9      55000 руб.  Менеджер по работе с клиентами, Pre-sale менед...   
10     55000 руб.              Менеджер / оператор интернет-магазина   
11     80000 руб.                               Технический писатель   
12     40000 руб.                                Инженер-програм

In [35]:
print(df[df['ЗП (руб)'].isna()])

Empty DataFrame
Columns: [ЗП, Ищет работу на должность:, Последнее/нынешнее место работы, Последняя/нынешняя должность, Обновление резюме, Авто, Образование, Пол, Возраст, Опыт работы (месяц), Город, Готовность к переезду, Готовность к командировкам, волонтерство, полная занятость, проектная работа, стажировка, частичная занятость, вахтовый метод, гибкий график, полный день, сменный график, удаленная работа, Дата, ЗП (руб)]
Index: []

[0 rows x 25 columns]


# Исследование зависимостей в данных

1. Постройте распределение признака **"Возраст"**. Опишите распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится возраст большинства соискателей? Есть ли аномалии для признака возраста, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [38]:
# ваш код здесь

ваши выводы по графику здесь

2. Постройте распределение признака **"Опыт работы (месяц)"**. Опишите данное распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится опыт работы большинства соискателей? Есть ли аномалии для признака опыта работы, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [41]:
# ваш код здесь

ваши выводы здесь

3. Постройте распределение признака **"ЗП (руб)"**. Опишите данное распределение, отвечая на следующие вопросы: каковы предельные значения признака, в каком примерном интервале находится заработная плата большинства соискателей? Есть ли аномалии для признака возраста? Обратите внимание на гигантские размеры желаемой заработной платы.
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*


In [44]:
# ваш код здесь

ваши выводы здесь

4. Постройте диаграмму, которая показывает зависимость **медианной** желаемой заработной платы (**"ЗП (руб)"**) от уровня образования (**"Образование"**). Используйте для диаграммы данные о резюме, где желаемая заработная плата меньше 1 млн рублей.
*Сделайте выводы по представленной диаграмме: для каких уровней образования наблюдаются наибольшие и наименьшие уровни желаемой заработной платы? Как вы считаете, важен ли признак уровня образования при прогнозировании заработной платы?*

In [47]:
# ваш код здесь

ваши выводы здесь

5. Постройте диаграмму, которая показывает распределение желаемой заработной платы (**"ЗП (руб)"**) в зависимости от города (**"Город"**). Используйте для диаграммы данные о резюме, где желая заработная плата меньше 1 млн рублей.
*Сделайте выводы по полученной диаграмме: как соотносятся медианные уровни желаемой заработной платы и их размах в городах? Как вы считаете, важен ли признак города при прогнозировании заработной платы?*

In [50]:
# ваш код здесь

ваши выводы здесь

6. Постройте **многоуровневую столбчатую диаграмму**, которая показывает зависимость медианной заработной платы (**"ЗП (руб)"**) от признаков **"Готовность к переезду"** и **"Готовность к командировкам"**. Проанализируйте график, сравнив уровень заработной платы в категориях.

In [53]:
# ваш код здесь

In [54]:
ваши выводы здесь

SyntaxError: invalid syntax (1506124366.py, line 1)

7. Постройте сводную таблицу, иллюстрирующую зависимость **медианной** желаемой заработной платы от возраста (**"Возраст"**) и образования (**"Образование"**). На полученной сводной таблице постройте **тепловую карту**. Проанализируйте тепловую карту, сравнив показатели внутри групп.

In [ ]:
# ваш код здесь

ваши выводы здесь

8. Постройте **диаграмму рассеяния**, показывающую зависимость опыта работы (**"Опыт работы (месяц)"**) от возраста (**"Возраст"**). Опыт работы переведите из месяцев в года, чтобы признаки были в едином масштабе. Постройте на графике дополнительно прямую, проходящую через точки (0, 0) и (100, 100). Данная прямая соответствует значениям, когда опыт работы равен возрасту человека. Точки, лежащие на этой прямой и выше нее - аномалии в наших данных (опыт работы больше либо равен возрасту соискателя)

In [ ]:
# ваш код здесь

ваши выводы здесь

**Дополнительные баллы**

Для получения 2 дополнительных баллов по разведывательному анализу постройте еще два любых содержательных графика или диаграммы, которые помогут проиллюстрировать влияние признаков/взаимосвязь между признаками/распределения признаков. Приведите выводы по ним. Желательно, чтобы в анализе участвовали признаки, которые мы создавали ранее в разделе "Преобразование данных".


In [ ]:
# ваш код здесь

ваши выводы здесь

# Очистка данных

1. Начнем с дубликатов в наших данных. Найдите **полные дубликаты** в таблице с резюме и удалите их. 

In [ ]:
# ваш код здесь

2. Займемся пропусками. Выведите информацию **о числе пропусков** в столбцах. 

In [ ]:
# ваш код здесь

3. Итак, у нас есть пропуски в 3ех столбцах: **"Опыт работы (месяц)"**, **"Последнее/нынешнее место работы"**, **"Последняя/нынешняя должность"**. Поступим следующим образом: удалите строки, где есть пропуск в столбцах с местом работы и должностью. Пропуски в столбце с опытом работы заполните **медианным** значением.

In [ ]:
# ваш код здесь

4. Мы добрались до ликвидации выбросов. Сначала очистим данные вручную. Удалите резюме, в которых указана заработная плата либо выше 1 млн. рублей, либо ниже 1 тыс. рублей.

In [ ]:
# ваш код здесь

5. В процессе разведывательного анализа мы обнаружили резюме, в которых **опыт работы в годах превышал возраст соискателя**. Найдите такие резюме и удалите их из данных


In [ ]:
# ваш код здесь

6. В результате анализа мы обнаружили потенциальные выбросы в признаке **"Возраст"**. Это оказались резюме людей чересчур преклонного возраста для поиска работы. Попробуйте построить распределение признака в **логарифмическом масштабе**. Добавьте к графику линии, отображающие **среднее и границы интервала метода трех сигм**. Напомним, сделать это можно с помощью метода axvline. Например, для построение линии среднего будет иметь вид:

`histplot.axvline(log_age.mean(), color='k', lw=2)`

В какую сторону асимметрично логарифмическое распределение? Напишите об этом в комментарии к графику.
Найдите выбросы с помощью метода z-отклонения и удалите их из данных, используйте логарифмический масштаб. Давайте сделаем послабление на **1 сигму** (возьмите 4 сигмы) в **правую сторону**.

Выведите таблицу с полученными выбросами и оцените, с каким возрастом соискатели попадают под категорию выбросов?

In [ ]:
# ваш код здесь

ваш коммментарий здесь